### DISCUSSION TITLE
Thanks for this!

### DISCUSSION THREAD

**Discussion by smashtup** [nirnachmani/Pixie-Plus-for-Home-Assistant#2] [General]
Cheers for your work on this. I have my setup now working through home assistant. For me, it was more about having a faster integration into HomeKit. Would still be nice to have direct-to Bluetooth.

Anyhoo, couple of observations/changes:

- for the pixie plus login it definitely helps to create a new one to use with HA only. Seems that one account can't be used between two locations without logging out of the other.
- When extracting the app key, install key, and Javascript key, I found Proxyman on the iPhone to work really well. Not very difficult at all to find things
- Added two new model numbers (2212, 2312) for the Gen 2 Pixie Switch and Dimmer. The commands you use still worked correctly for these models 

The biggest gotcha though, I had all my lights working perfectly except for one. After reverse engineering your commands with those that come out of the pixie plus app (using Proxyman and making sure you disable Bluetooth), it appears that you may have made the assumption that the last two hex numbers in the mac address are what are needed in the command. It appears that these hex numbers always match the Device ID (converted to decimal obviously) but this is not always the case. An alternative is just to convert the device id to hex and use that instead in your light command.

Anyway, thanks again.

**Comment by nirnachmani** [comment]
Hi Michael,

I am glad my integration is of use and is overall working! 

I agree about the need to create a new pixie plus login and not to use the same one you use on your mobile, as they can't both be logged in at the same time (different sessionID). However, if you use a virtual phone to retrieve the app, install and java keys, it creates new credentials because it is a different phone, even if you use the same username/password.

It's great to know you can use Proxyman to retrieve those values - I'm not familiar with it. However, because you use it on your existing phone/account you will be using the same credentials (app, install and java keys) in HA, which won't work, hence, the need to create a new account when you do it that way. I'll add this way of doing it to the instruction when I have a moment.  

Is the 2212 the switch and the 2312 a dimmer? 

Thank you for letting me know about the Device ID issue. I did make the assumption that they use the last two numbers of the mac address for the commands because this seemed to be the case with all my devices. I have been wondering why they are not using the Device ID and didn't think about the possibility that they do, but convert it to hex! They probably base the device ID on the mac, except in cases where there are two devices with the same last two mac hex numbers. I will change the code to use the device ID (converted to hex). Thanks for realising this!

Cheers

**Comment by nirnachmani** [comment]
Hi Michael,

Can you please send me the names and model numbers (as on the box of the product) for 2212 and 2312 so I can add them formally?

I changed the code so the device ID is used for commands. 

Thanks.

**Reply by smashtup** [reply]
Apologies for the delay:

Basically it's just the Gen 2 versions of what you've already got for switch and dimmer

    "2212": "Smart Switch G2 - SWL350BT",
    "2312": "Smart Dimmer G2 - SDD350BT",

I probably should've versioned my changes so I could've submitted a pull request to your code.

So far I've fixed:
- added new devices as above
- switched to device id from mac
- fixed and interesting issue where home assistant would reset dimmers back to max brightness when turning on lights.

Nothing clean enough to send back to you though :P

Cheers,

**Reply by nirnachmani** [reply]
No problem! I'll add the devices to the code and already changed so it is using the device id. 

With the dimmers being set to max, did you change it so the last brightness is used when turning on? I think the problem is the way I send the 'on' command, it sets it to max. Should probably read the brightness parameter and use that instead.

Thanks

**Reply by nirnachmani** [reply]
I fixed the brightness being set to max when turned on. The problem is that the brightness parameter is updating from the light's current state as reported by Pixie Plus and when the light turns off, the brightness parameter is set to 0. When the light is turned back on, it doesn't remember the last brightness. I added a property called last_brightness that stores the brightness when the light is turned off and is then used when it turns on again. I also added last_rgb_color so it remembers the color when turned off. Unfortunately, the last_brightness won't work after a HA restart and the light will be again set to max when turned on. Restoring the brightness won't work because if the light was off when AH restart occurred, the last brightness would be 0.

**Reply by smashtup** [reply]
Exactly what I discovered. 

To fix restoring brightness I use the switch command for on (ie. no brightness hex etc) in which case the dimmer will restore to previous brightness by itself. Condition for this is when turning on the light if the current brightness is 0 then send a switch on command instead. Works a treat!

**Reply by nirnachmani** [reply]
Sounds good. Then it is not affected if HA is restated. Can you send me the code? 

The problem will be with light with color and dimming. Sending 'on' and not brightness value will mean that none of the color settings will take effect. May need to make sure that light has dimming but not colors before sending 'on' command instead of brightness value.

**Reply by smashtup** [reply]
Interesting that the lights could remember brightness but not colour. Definitely need a better solution for colour + brightness restoring.

Code is exactly as you described: 

```
        if self._model_no in has_dimming and self._brightness > 0:
            brightness_hex = (f'{int(self._brightness):x}').zfill(2)
        else:
            brightness_hex = "on"
```

Also, made sure that brightness is always stored as a value from 0..255. The only place where a conversion is then needed is the websocket update on brightness since it's decimal instead of hex. Definitely a WTH moment from Pixie.

**Reply by nirnachmani** [reply]
Sorry, you are right. If I use the 'on' command on RGB lights it remembers the last state (brightness and colour). I'll change the code accordingly as per your code.

With the brightness value, I'm not sure I'm following you. Pixie Plus provides a number between 1-100 and HA provides 0-255. For the command I need the 0-255 value, but in hex. I store the brightness value in self._brightness as 0-255, converting it from 1-100 to 0-255 every time I get it from Pixie Plus. I convert it to hex when sending the command with hex()[2:].zfill(2). You are saying I don't need to convert it to hex?  Or that because it is 0-255 it needs another conversion? I know that two hex digits represent values of 0-255 but I still need it in hex format, no? How is self_brightness changed by using the f'{int(self._brightness):x}'? I understand that the f is for formatting but what does it do in this context? and what does the :x does?  Sorry for asking what are probably trivial questions but I couldn't find an answer on google. 

It would be good if you update my code with those changes, if you have the time.   

Cheers

**Comment by smashtup** [comment]
My bad on the brightness front. Pretty sure we’re on the same page.

Brightness setting, whether it be sending it to pixie or storing it in HA, it’s always a value from 0..255. When sending to pixie it needs to converted to hexidecimal first. So there should be no conversions to a percentage brightness ie. 0..100

The weird thing is as you pointed out, the websocket update that comes from pixie reports a decimal percentage from 0..100. Hence it needs to be converted to percentage of 255 before storing in HA.

The format string you see is just converting to hex without 0x. Probably can get rid of zfill as well looking at some of the format string examples. https://bobbyhadz.com/blog/python-get-hex-representation-without-0x#:~:text=Use%20a%20formatted%20string%20literal,hex%20without%20the%200x%20prefix.

Michael Coleman

**Comment by robothomessydney** [comment]
Hey Guys,

Hope you are well. Thank for your hard work to get this integration done. Looks like you have done an amazing job. 
We are currently trying to implement your integration but we are very stuck at obtaining this Javascript key. We are using proxyman on an IOS device. We can clearly see the Application ID and Installer ID but cannot find the Javscript key. Would you please be able to assist us? In either where to obtain it using Proxyman, or maybe what we are looking for, like an example? 
Michael Coleman you seem like the one who we could really help. Thanks in advance!
